Vulkan backend

sheaf opened the pull request:
I have no idea what I'm doing, so I would appreciate if you could take a look and see if this is even remotely sensible. Thanks.

ocharles commented on src/DearImGui/Vulkan/Types.hs:
```diff
@@ -0,0 +1,38 @@
+  , ( C.TypeName "VkSampleCountFlagBits", [t|Vulkan.SampleCountFlagBits|] )
+  ]
+
+addVulkanTypesTable :: C.Context -> C.Context
+addVulkanTypesTable ctx@( C.Context { C.ctxTypesTable } ) = ctx { C.ctxTypesTable = ctxTypesTable <> vulkanTypesTable }
```
I think what you usually do with `inline-c` is to just provide a `C.Context`, and then you can use its `Monoid` instance to combine multiple context together. So this would be:

```suggestion
vulkanContext :: C.Context
vulkanContext = mempty { C.ctxTypesTable = vulkanTypesTable }
```

ocharles commented on src/DearImGui/Vulkan.hs:
```diff
@@ -0,0 +1,163 @@
+  ( addVulkanTypesTable )
+
+
+C.context ( addVulkanTypesTable (Cpp.cppCtx <> C.funCtx) )
```
```suggestion
C.context ( Cpp.cppCtx <> C.funCtx <> vulkanContext )
```

ocharles reviewed (commented):
@sheaf I can't see anything obviously wrong, and only have a suggestion about `Context`.

sheaf commented:
Currently failing with the following error:

```
(debug) Recording one-shot commands
<interactive>: internal error: stg_ap_p_ret
    (GHC version 8.10.3 for x86_64_unknown_mingw32)
    Please report this as a GHC bug:  https://www.haskell.org/ghc/reportabug
```

This might indicate an issue with `vulkanCreateFontsTexture` as that's the next thing called.

sheaf commented:
OK, as we discovered the issue was that the function pointer for the `checkResult` function was being freed too early. I changed things around to have a `withVulkan` function that frees the function pointer manually at the end.    

Should be good to go now!

ocharles commented on dear-imgui.cabal:
```diff
@@ -97,3 +129,34 @@ executable readme
+    , base
+        >= 4.13 && < 4.16
+    , bytestring
+        >= 0.10.12.0 && < 0.12
```
I had to relax this locally to get this to build:

```suggestion
        >= 0.10.10.0 && < 0.12
```

ocharles commented on src/DearImGui/Vulkan.hs:
```diff
@@ -0,0 +1,162 @@
+    withCallbacks f = case mbAllocator of
+      Nothing        -> f nullPtr
+      Just callbacks -> alloca ( \ ptr -> poke ptr callbacks *> f ptr )
+  checkResultFunPtr <- liftIO $ $( C.mkFunPtr [t| Vulkan.Result -> IO () |] ) checkResult
+  res <- liftIO $ withCallbacks \ callbacksPtr ->
+      [C.block| bool {
+        ImGui_ImplVulkan_InitInfo initInfo;
+        VkInstance instance = { $( VkInstance_T* instancePtr ) };
+        initInfo.Instance = instance;
+        VkPhysicalDevice physicalDevice = { $( VkPhysicalDevice_T* physicalDevicePtr ) };
+        initInfo.PhysicalDevice = physicalDevice;
+        VkDevice device = { $( VkDevice_T* devicePtr ) };
+        initInfo.Device = device;
+        initInfo.QueueFamily = $(uint32_t queueFamily);
+        VkQueue queue = { $( VkQueue_T* queuePtr ) };
+        initInfo.Queue = queue;
+        initInfo.PipelineCache = $(VkPipelineCache pipelineCache);
+        initInfo.DescriptorPool = $(VkDescriptorPool descriptorPool);
+        initInfo.Subpass = $(uint32_t subpass);
+        initInfo.MinImageCount = $(uint32_t minImageCount);
+        initInfo.ImageCount = $(uint32_t imageCount);
+        initInfo.MSAASamples = $(VkSampleCountFlagBits msaaSamples);
+        initInfo.Allocator = $(VkAllocationCallbacks* callbacksPtr);
+        initInfo.CheckVkResultFn = $( void (*checkResultFunPtr)(VkResult) );
+        return ImGui_ImplVulkan_Init(&initInfo, $(VkRenderPass renderPass) );
+      }|]
+  a <- action ( res /= 0 )
+  liftIO do
+    [C.exp| void { ImGui_ImplVulkan_Shutdown(); } |]
+    freeHaskellFunPtr checkResultFunPtr
```
Should this be changed to be `bracket`? If `action` crashes, for example, we still want to call `freeHaskellFunPtr`

sheaf replied:
OK... I'm not too familiar with the details, do we need an uninterruptibleMask, or is it OK to just use `Control.Exception.bracket`? And what if ImGui_ImplVulkan_ShutDown crashes?

ocharles replied:
>  And what if ImGui_ImplVulkan_ShutDown crashes?

I think this is the crux of the question as to whether you should mask. I think just `bracket` would be fine here (though it's not a particular informed judgement). But it at least catches the case where an exception is thrown from `action`.

sheaf replied:
I realised I can't use bracket as is because of the `MonadIO` action type. Should I use `UnliftIO`'s bracket?

ocharles replied:
Ah, good point. Yea, let's go with `UnliftIO.bracket`.

ocharles reviewed (approved):
This is excellent work - thank you so much for taking the time to dive into this! I have no objections to merging this as is. If you want to do more work on the example code first that's fine, just give me the heads up when you want to merge.

sheaf commented:
- Added bracket to `withVulkan` function.
- Fixed example code: enable window resizing, and properly re-create swapchain (and associated resources) on the fly on resize events.

You should check you can indeed resize the window on your end, and we should be good to go for merging this!
